pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/NoDictionaryOnHeapDictionaryJointRule.java

```diff
@@ -0,0 +1,245 @@
+    AtomicDouble totalWeight = new AtomicDouble(0);
+
+    //**********No dictionary recommendation*******/
+    Set<String> noDictCols = new HashSet<>(_inputManager.getColNameToIntMap().keySet());
+
+    //Exclude cols with index
+    noDictCols.removeAll(_outputManager.getIndexConfig().getInvertedIndexColumns());
+    noDictCols.removeAll(_outputManager.getIndexConfig().getSortedColumn());
+    // TODO: Remove this after range index is implemented for no-dictionary
+    noDictCols.removeAll(_outputManager.getIndexConfig().getRangeIndexColumns());
+    LOGGER.debug("noDictCols {}", noDictCols);
+    //Find out columns used in filter&groupby and selection and corresponding frequencies
+    _inputManager.getQueryWeightMap().forEach((query, weight) -> {
+      parseQuery(query, weight, filterGroupByWeights, selectionWeights);
+      totalWeight.addAndGet(weight);
+    });
+
+    //Add dictionary on columns used in filter&groupby , with frequency > threshold
+    for (int i = 0; i < numCols; i++) {
+      double filterGroupByFreq = filterGroupByWeights[i] / totalWeight.get();
+      if (filterGroupByFreq > _params.THRESHOLD_MIN_FILTER_FREQ_DICTIONARY) {
+        noDictCols.remove(_inputManager.intToColName(i));
+      }
+    }
+
+    LOGGER.debug("filterGroupByWeights {}, selectionWeights{}, totalWeight{} ", filterGroupByWeights, selectionWeights,
+        totalWeight);
+    LOGGER.debug("noDictCols {}", noDictCols);
+
+    for (int i = 0; i < numCols; i++) {
+      // No dictionary on columns frequently used in selection
+      double selectionFreq = selectionWeights[i] / totalWeight.get();
+      if (selectionFreq > _params.THRESHOLD_MAX_SELECTION_FREQ_DICTIONARY) {
+        continue;
+      }
+
+      // Add dictionary on columns NOT frequently used in selection
+      // AND can save storage > threshold
+      String colName = _inputManager.intToColName(i);
+      double noDictSize;
+      double withDictSize;
+      long colDataSizeWithoutDictionary = _inputManager.getColDataSizeWithoutDictionary(colName);
+      double numValuesPerEntry = _inputManager.getNumValuesPerEntry(colName);
+      int bitCompressedDataSize = _inputManager.getBitCompressedDataSize(colName);
+      long dictionarySize = _inputManager.getDictionarySize(colName);
+      double cardinality = _inputManager.getCardinality(colName);
+      long numRecordsPerPush = _inputManager.getNumRecordsPerPush();
+      LOGGER.debug("colDataSizeWithoutDictionary {}", colDataSizeWithoutDictionary);
+      LOGGER.debug("bitCompressedDataSize {}", bitCompressedDataSize);
+      LOGGER.debug("dictionarySize {}", dictionarySize);
+      LOGGER.debug("numValuesPerEntry {}", numValuesPerEntry);
+
+      if (_inputManager.getTableType().equalsIgnoreCase(REALTIME)) {
+        //TODO: improve this estimation
+        noDictSize = // size of one segment flushed ith no dictionary
+            colDataSizeWithoutDictionary * numValuesPerEntry * _params.SEGMENT_FLUSH_TIME;
```

Comment:
We can simplify this by not bringing numValuesPerEntry into the equation. numValuesPerEntry is applicable to MV columns right? Pinot currently doesn't support raw MV columns so they are **always** dictionary encoded

Reply:
We should add a TODO though -- stating to enhance this to consider MV columns as noDictionary in the future when Pinot supports that

Reply:
Done. Thanks for bring out this!